Update chowder.py

VolodymyrChapman opened the pull request:
Add metadata_cols argument to Chowder class to allow for input of data with preceding columns (for example, magnification) to be ignored by the chowder model. 
Default of 3 assumes magnification, start_x and start_y columns preceding data to be fitted / inferred on, in keeping with current behaviour of 3 preceding columns - i.e. default settings will not break behaviour of existing Chowder models / code.
Testing on fresh conda environment with modified Chowder model:

![image](https://github.com/owkin/HistoSSLscaling/assets/44582194/d39717d2-9778-48cb-ae4c-ff6c15f74a65)

VolodymyrChapman committed:
Update chowder.py

Add metadata columns argument to Chowder class

jbschiratti commented:
Thanks @VolodymyrChapman for initiating this PR.

To keep a consistent behavior, can you also add the `metadata_cols` parameter in other [MIL models](https://github.com/owkin/HistoSSLscaling/tree/main/rl_benchmarks/models/slide_models)?

VolodymyrChapman commented:
@jbschiratti Of course - will do :-)

VolodymyrChapman committed:
Update abmil.py

Add metadata_cols variable for preceding columns to be ignored

VolodymyrChapman committed:
Update dsmil.py

Add metadata_cols argument for preceding columns to be ignored in training / inference

VolodymyrChapman committed:
Update meanpool.py

Add metadata_cols argument for preceding data columns to ignore

VolodymyrChapman committed:
Update transmil.py

Add metadata_cols argument to TransMIL to ignore preceding metadata columns in training / inference

VolodymyrChapman committed:
Update hiptmil.py

Added metadata_cols argument to HIPTMIL class, to set number of preceding metadata columns to ignore for training / inference

VolodymyrChapman committed:
Update hiptmil.py

Added metadata_cols argument in super transformer class

jbschiratti commented:
Thanks @VolodymyrChapman!

Although there's no CI (yet) on this repository, can you, please, update the unit tests and make sure they pass locally?

VolodymyrChapman committed:
Update test_slide_models.py

Update slide model tests with metadata_col argument

VolodymyrChapman committed:
Update test_slide_models.py

Correct wrong variable name for metadata_cols

VolodymyrChapman committed:
linting

Linting with black, as per linting tests

jbschiratti commented:
@VolodymyrChapman Looks like you pushed some unnecessary `.pyc` files by mistake. Can you please remove them from this PR?

VolodymyrChapman commented:
@jbschiratti Of course :-)
The metadata_cols variable has been added to all MIL models, as per the Chowder model.
The variable has also been added to all models in the tests/models/test_slide_models.py file.
Linting was performed with black using the specifications outlined in dev_tools/linting.sh
Local output of black checks and unittest for just the MIL models (Ubuntu 22.04 in a fresh conda env):
![image](https://github.com/owkin/HistoSSLscaling/assets/44582194/38f7cbe2-2fd0-4c9e-84e8-7d8d5b866311)

Are there any other tests / checks that you would require?

jbschiratti commented:
@VolodymyrChapman Once you've removed the `.pyc` files, this will be good to go. 

I'll let @afilt take over as he's the code owner on this repository.

VolodymyrChapman committed:
removing pyc files

VolodymyrChapman committed:
Deleting pyc files

VolodymyrChapman commented:
@jbschiratti  Thank you for spotting - apologies for the spam files!
I think they've all been removed now.

afilt commented on rl_benchmarks/models/slide_models/abmil.py:
```diff
@@ -44,6 +44,10 @@ class ABMIL(nn.Module):
+    metadata_cols: int = 3
+        Number of metadata columns (for example, magnification, patch start
+        coordinates etc.) at the start of input data. Default of 3 assumes
+        magnification, patch start x and patch start y.
```
Minor rephrasing, could you replace by smgth more specific: "Default of 3 assumes the deep zoom level (corresponding to a given magnification), x and y coordinates of the patch." ? Thanks a lot !

VolodymyrChapman replied:
Hi @afilt - my pleasure.
Thank you for the feedback - I believe this has been actioned now. Please let me know whether the explanations are satisfactory now. :-)

afilt commented on rl_benchmarks/models/slide_models/chowder.py:
```diff
@@ -54,6 +54,10 @@ class Chowder(nn.Module):
+    metadata_cols: int = 3
+        Number of metadata columns (for example, magnification, patch start
+        coordinates etc.) at the start of input data. Default of 3 assumes
+        magnification, patch start x and patch start y.
```
Same applies here.

VolodymyrChapman replied:
As previous comment :-)

afilt commented on rl_benchmarks/models/slide_models/abmil.py:
```diff
@@ -138,7 +145,7 @@ def forward(
             (B, OUT_FEATURES), (B, N_TILES)
         """
-        tiles_emb = self.tiles_emb(features[..., 3:], mask)
+        tiles_emb = self.tiles_emb(features[..., self.metadata_cols :], mask)
```
Could you please remove the blank space between `self.metadata_cols` and `:` ? You can apply black formatting using `black --line-length=79 --check --diff rl_benchmarks/models/slide_models/`. Thanks ! Same applies for the other files.

VolodymyrChapman replied:
Blank spaces manually removed :-)
FYI - I believe that it is black formatting that added this space as the --diff output suggests addition of it:
![image](https://github.com/owkin/HistoSSLscaling/assets/44582194/990a6e1e-a53e-4945-b9a7-cdaaad6033ad)

afilt reviewed (commented):
Thanks a lot @VolodymyrChapman for adding this parameter to the MIL models. I left 2 minor comments that would require changes imo. Thanks @jbschiratti for the follow-up !

VolodymyrChapman committed:
Detailed metadata_cols explan. and whitespace

VolodymyrChapman committed:
Revert Camelyon16.py

VolodymyrChapman committed:
Remove accidental whitespace

afilt reviewed (approved):
Looks good to me, thanks a lot @VolodymyrChapman 🚀 Can you merge and delete your branch ? Thanks.

VolodymyrChapman commented:
@afilt Thank you! I don't think I have necessary write access to perform the merge as the button is missing from my view:
![image](https://github.com/owkin/HistoSSLscaling/assets/44582194/0e7504f6-10ba-41d6-98c1-01888116ef55)
Please let me know if the merge is successful from your end so I can delete the fork :-)

afilt commented:
You're right :-) Thanks a lot @VolodymyrChapman !

VolodymyrChapman commented:
My pleasure - happy to help :-)
